README.md
![Downloads](https://pypip.in/d/trepan/badge.png) [![Build Status](https://travis-ci.org/rocky/python2-trepan.svg)](https://travis-ci.org/rocky/python2-trepan) ![Supported Python versions](https://pypip.in/py_versions/trepan/badge.svg)

This is a rewrite of the Python debugger, [pydb](http://bashdb.sf.net/pydb), itself a derivative of the stock Python debugger *pdb*.

See the [wiki](https://github.com/rocky/python2-trepan/wiki) for
more information.

Works with Python 2.6 and 2.7. It might also work with Python 2.5.
See [this](https://code.google.com/p/python3-trepan/) for the same thing
that works with Python 3. Use *pydb* for Python 2.5 and earlier.


NEWS
trepan 0.8.7 2018-20-27 9 x 7 release

- Command doc changes

trepan 0.8.7 2018-04-16  (no 0.8.5 or 0.8.6)

- gdb "backtrace" negative count documented and supported
- add -d, --deparse and --source options on backtrace command
- DRY uncompyle and deparsing code. Use newer API
- expand source-code line-number width to 4 places
- reinstante deparse

trepan 0.8.4 2018-02-5

- Add "break!" to set a breakpoint whether trepan thinks
  there is a breakpoint
- Cache deparse info
- "set style" improvements
- Handle PyPy better
- "info locals *" and "info globals *" will list names
   in columnar form omitting values

trepan 0.8.3 2018-01-21

- Add break! and b! aliases to force setting a breakpoint on a line
- Update rst
- print p and pr mean the same thing

trepan 0.8.2 2017-12-10

- Misc little fixes
- Go over rst for revised list commands

trepan 0.8.1 2017-10-28

Fix botched release

- Packaging woes; was missing modules
- Correct documentation

trepan 0.8.0 2017-10-27

User incompatibilty alert!

We have redone location, breakpoint, and list-range parsing.

This release needs an explanation.

Code to parse commands like "list" and "breakpoint" were ugly and hard
to maintain. I hit a the tipping point in adding flexibility to the
"disassemble" command which allows address in addition to the usual
other kinds of locations.

Since version 0.6.3 when the "deparse" command was added, there has
been a hidden dependency of the the Earley parser.

That is now used to simplify parsing concepts like location,
list range, or breakpoint position.

In the process, I've gone over the syntax to make things more gdb
compatible. But of course gdb is also a moving target, so its syntax
has been extended and gotten more complicated as well.

In the olden days, I was sad that debuggers didn't follow someone
else's syntax but instead invented their own, sometimes incompatible
with gdb. Nowadays though it is a Herculean feat to come close to
matching gdb's syntax. Sigh.

The other problem with matching gdb's syntax is that debugging Python
is enherently different from debugging a compiled language with object
files. Python's language model just isn't the same as C's.

So there are deviations. The biggest change that I suspect will impact
users is that function names in locations in this debugger needs a
trailing "()" to mark it as a function. This was not needed in
previous versions and it isn't needed in gdb.

Not implimented in our notion of location are things that feel
compiled-language object-file-ish. Specifying the function name inside
an object file, isn't the way Python (or most dynamic languages)
do things.  Instead you list the method/function inside a class or
module. And we allow this to be done off of variables and variables
holding instance methods.

Some things like ending at an address is not implemented as going
backwards in variable-length bytecode is a bit of work. Other things
of dubious merit I've omitted. The flexibility that is there
is probably overkill.

Speaking of reduced flexibility. Now with parser in place we no longer
support expressions as numbers in list commands. It's not in gdb and
I have a feeling  that too is overkill.


trepan 0.7.8 2017-07-10

- Release for updated xdis and botched Rst
- PyPy 2.x tolerance
- Small error message improvements

trepan 0.7.7 2017-07-10

- improve remote debugging:
  * scans for open ports
  * allow a socket to be passed in
  * start Celery remote debugging
- add deval, deval? commands: deparsed eval and eval?
- in python/shell don't go into post-mortem debugger on exception
- find source or decompile when bytecode is given
- Handle position in the presence of eval() and exec()
- show code via xdis when disassembling an entire function.
- improved terminal background detection respecting
  environment variables: DARK_BG, COLORFGBG
- deparse improvements:
  * do deparse offset fuzzing
  * show asm listing for opcode
  * add -o | --offset to indicate where to start deparsing from

trepan python-2.4-0.7.6 2017-07-01

- Retrofit back to Python 2.4

trepan 0.7.6 2017-06-03
- Fix botched release and sync version numbers with trepan3k
- Corrected setup.py code, cheking and better error messaging
`- better stack trace position in the presense of eval() and exec()
- position in the presence of eval() and exec() via uncompyle6; this needs
  uncompyle6 0.11.0 or greater now
- better deparse command output shows grammar sysmbol on parent,
  and full disassembly line for instruction
- deparse offset fuzzing
- fix bugs in deparser option processing
- in disassembly of functions show function header
- go over docs info.rst, pc.rst

trepan 0.7.4 2017-06-03 Marilyn Frankel
- Fix botched release
  * add rst files
  * go over docs even more
  * fix packaging adminstrivia which botched release

trepan 0.7.3 2017-06-03
- Package name is now trepan2!
- Add version test in setup to avoid possible install from Python 3
- Resync version number with Python 3 trepan3k
- Add "info pc" command
- Show list of styles with "set style" (no style given)
- Add MS Windows kill with Python 2.6 tolerance
- Bugs in remote debugging tcpfns.py: handle "kill" better client.py: remove extra "\n"s
- Document "set substitute"
- Document "set style"
- "deparse" changes:
  * add --offset/-O option to show exact deparsable offsets
  * allow fuzzy offset deparsing via uncompyle6 deparse_around_offset
- "list" changes:
  * deparse when no source is available (e.g. on an exec or eval string)
- "set highlight" changes:
  * clear source-code cache after setting highlight
- "show display changes"
  * show display expression after setting it
- Update MANIFEST.in which should provide more reliable packaging
- "disassemble" changes:
  * better output using routines from xdis package
- force deparse improvements by bumping uncompyle6 minimum version

trepan 0.6.5 2016-07-26

- PyPy tolerance
- Add deparse options --parent --AST, --offset
- Use deparse bytecode to get source if we can't find it

__pkginfo__.py
# Copyright (C) 2008-2010, 2013-2018 Rocky Bernstein <rocky@gnu.org>
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""Debugger packaging information"""

# To the extent possible we make this file look more like a
# configuration file rather than code like setup.py. I find putting
# configuration stuff in the middle of a function call in setup.py,
# which for example requires commas in between parameters, is a little
# less elegant than having it here with reduced code, albeit there
# still is some room for improvement.

# Things that change more often go here.
copyright   = """
Copyright (C) 2008-2010, 2013-2018 Rocky Bernstein <rocky@gnu.org>.
"""

classifiers =  ['Development Status :: 5 - Production/Stable',
                'Environment :: Console',
                'Intended Audience :: Developers',
                'License :: OSI Approved :: GNU General Public License v3 (GPLv3)',
                'Operating System :: OS Independent',
                'Programming Language :: Python',
                'Topic :: Software Development :: Debuggers',
                'Topic :: Software Development :: Libraries :: Python Modules',
                'Programming Language :: Python :: 2.4',
                'Programming Language :: Python :: 2.5',
                'Programming Language :: Python :: 2.6',
                'Programming Language :: Python :: 2.7',
                ]

# The rest in alphabetic order
author             = "Rocky Bernstein"
author_email       = "rocky@gnu.org"
ftp_url            = None
install_requires   = ['columnize >= 0.3.9',
                      'nose>=1.0.0, <= 1.3.7',
                      'pyficache >= 1.0.0',
                      'pygments  == 1.4',
                      'spark_parser >= 1.8.5, <1.9.0',
                      'spark_parser >= 1.8.7, <1.9.0',
                      'uncompyle6 >= 3.2.4',
                      'tracer >= 0.3.2',
                      'unittest2',
                      'xdis >= 3.8.8',
                      ]
license            = 'GPL3'
mailing_list       = 'python-debugger@googlegroups.com'
modname            = 'trepan2'
packages = [
    'trepan',
    'trepan.bwprocessor',
    'trepan.interfaces',
    'trepan.inout',
    'trepan.lib',
    'trepan.processor',
    'trepan.processor.command',
#   'trepan.processor.command.ipython_magic',
    'trepan.processor.command.info_subcmd',
    'trepan.processor.command.set_subcmd',
    'trepan.processor.command.show_subcmd',
    'trepan.processor.parse',
]
py_modules         = None
short_desc         = 'GDB-like Python Debugger in the Trepan family'

import os

def get_srcdir():
    filename = os.path.normcase(os.path.dirname(os.path.abspath(__file__)))
    return os.path.realpath(filename)

# VERSION.py sets variable VERSION.
ns = {}
exec(open(os.path.join(get_srcdir(), 'trepan', 'version.py')).read(), ns)
version            = ns['VERSION']
web                = 'http://github.com/rocky/python2-trepan/'

# VERSION.py sets variable VERSION.
ns = {}
exec(open(os.path.join(get_srcdir(), 'trepan', 'version.py')).read(), ns)
version            = ns['VERSION']

# tracebacks in zip files are funky and not debuggable
zip_safe = False


def read(*rnames):
    return open(os.path.join(os.path.dirname(__file__), *rnames)).read()

long_description   = ( read("README.rst") + '\n' )
